accepted/single-file/design.md

```diff
@@ -0,0 +1,281 @@
+* External tools like [Fody](https://github.com/Fody/Fody) that use  `AfterBuild`/`AfterPublish` targets may need to adapt to expect the significantly different output generated by publishing to a single file. The goal in this case is to provide sufficient documentation and guidance.
+
+### The Host
+
+On Startup, the AppHost checks if it has embedded files. If so, it 
+
+* Processes configuration files such as `app.deps.json` , `app.runtimeconfig.json` from the embedded files, instead of looking for them on the disk next to the app.
+* Sets up the data-structures to locate pure managed assemblies within the bundle so that they can be loaded on-demand.
+* Extracts the other files (native libraries, ready-to-run compiled assemblies, and other data files to an *install-location*). Extraction of files is discussed in detail in this  [document](extract.md). 
+* Communicates the information about bundled and extracted files to the runtime, as explained in the section on dependency resolution.
+
+#### Dependency Resolution
+
+An app may choose to only embed some files (ex: due to licensing restrictions) and expect to pickup other dependencies from application-launch directory, nuget packages, etc. In order to resolve assemblies and native libraries, the embedded resources are probed first, followed by other probing paths. 
+
+The communication between the host and the runtime will continue to be through properties such as `TRUSTED_PLATFORM_ASSEMBLIES`, `NATIVE_DLL_SEARCH_DIRECTORIES`, etc.  These properties will be populated with entries from the bundled files (using a special marker) and the extracted files.
+
+For example, if the TPA without single-exe publish is:
+`~/app/il.dll`; `~/app/r2r.dll`; `/usr/local/nuget/dep.dll`
+
+Where
+
+* `il.dll` is a pure managed assembly that is loaded directly from the bundle
+* `r2r.dll` is a ready-to-run assembly that is bundled and extract out to disk
+* `dep.dll` is an assembly that is not bundled with the app,
+
+the TPA with single-exe publish will be (assuming `?` is the bundle marker):
+`?/il.dll`; `/tmp/.net/e53xf3/r2r.dll`; `/usr/local/nuget/dep.dll`
+
+## Testing
+
+* Unit Tests to achieve good coverage on apps using managed, ready-to-run, native code 
+* Tests to verify that framework-dependent and self-contained apps can be published as single file.
+* Tests for ensure that every app model template supported by .NET Core (console, wpf, winforms, web, etc.) can be published as a single file.
+* A subset of CoreCLR Tests
+* End-to-End testing on real world apps such as Roslyn, MusicStore
+* *Measurements*: Publish size and run-time (first run, subsequent runs) for HelloWorld, Roslyn and MusicStore.
+
+## Further Work
+
+#### Bundler Optimizations
+
+Since all the files of an app published as a single-file live together, we can perform the following optimizations
+
+- R2R compile the app and all of its dependent assemblies in a single version-bubble
```

Comment:
Is this at odds with the other goal of "Able to generate cross-platform bundles (ex: publish a single-file for Linux target from Windows)"? My understanding is that at present, we do not support "cross-crossgen", correct?

Reply:
Thanks for pointing that out @nguerrera. Yes, we can generate cross-platform bundles without any new cross-compilation optimizations, until cross-crossgen is available.

Reply:
@swaroop-sridhar is there a "not" missing from the second sentence? Is it that we cannot until we have cross-crossgen?

Reply:
Yes that's a typo, sorry.